### PR TITLE
Implement retry logic for List() Operations

### PR DESCRIPTION
**Pull request by spencerhance**
List() calls will retry until successful or the context is done.  PageToken()
is used to avoid re-trying successful pages.

This should resolve https://github.com/GoogleCloudPlatform/k8s-cloud-provider/issues/29

Up for review but hold off on merging since I'm trying to figure out a better way to test this.

### CODE REVIEW

**Comment by spencerhance** [comment]
@bowei Can you review this since you're the most familiar with this code?  Also if you have any suggestions on unit testing that would be appreciated; we don't actually run any of the gen functions in the unit tests.

**Comment by bertinatto** [comment]
@spencerhance, @bowei : Thank you for looking into this!

If you think this is the way to go, I can help testing this patch in our CI for a couple of weeks.

**Comment by BenTheElder** [comment]
[verifying that prow webhooks work]
/woof

**Comment by bertinatto** [comment]
Hi @spencerhance, @bowei. Do you have any thoughts/concerns on this PR?

File: pkg/cloud/gen/main.go [hunk 1]
```diff
@@ -728,15 +728,33 @@ func (g *{{.GCEWrapType}}) List(ctx context.Context, zone string, fl *filter.F)
...
+		case <-ctx.Done():
+			klog.V(5).Infof("{{.GCEWrapType}} call.Pages(%v, _) not completed, poll count = %d, ctx.Err = %v", ctx, retryCount, ctx.Err())
+			return nil, ctx.Err()
+		default:
+			// ctx is not canceled, continue immediately
+		}
+		if nextPageToken != "" {
+			call.PageToken(nextPageToken)
+		}
+		if err := call.Pages(ctx, f); err != nil {
```
Comment (bowei) [inline pkg/cloud/gen/main.go:751]: Are there different kinds of errors we want to handle differently?

Comment (spencerhance) [reply pkg/cloud/gen/main.go:751]: I don't think so, are there any specific response codes you think we should handle?  We don't handle response codes individually in the other functions AFAIK

**Comment by bertinatto** [comment]
@bowei @spencerhance do you need any help testing this?

File: pkg/cloud/gen/main.go (hunk 1 above)
Comment (bertinatto) [reply pkg/cloud/gen/main.go:751]: @bowei @spencerhance please let me know if you need any help with this! We're seeing some flakes in our CI that I think would be solved with this patch.

**Comment by spencerhance** [comment]
@bowei Is there anything else I need to do to close this out?

**Comment by bowei** [comment]
Can we address the question about the kinds of errors we would or wouldn't want to retry? Also, I get concerned that current calls to List() that return with an error when called with a no deadline Context will now just hang.

**Comment by bowei** [comment]
Spencer can you update or abandon this PR?

**Comment by spencerhance** [comment]
Closing for now since it's somewhat of a high risk / low reward PR - callers can always just wrap List() in a retry loop.

**Pull request closed without merging**